[Core] Add a random suffix to frontend-provided request IDs

markmc opened the pull request:
Since #9550 and #10968 we support client's supplying a custom request ID. The motivation for this is that it can be very helpful when you need to correlate vLLM logs with logs of a related service.

Since the request ID is used ubiquitously across vLLM as a unique key, it obviously is problematic if we ever have multiple in-flight requests using the same client-provided request ID.

We saw this happening recently when `vllm serve bench` started including a request ID and the request IDs from multiple concurrent instances caused collisions. See #27723

We try to guard against request ID collisions currently in the frontend in OutputProcessor:

```
    def add_request(...):
        if request_id in self.request_states:
            raise ValueError(f"Request id {request_id} already running.")
```

however, this is not always effective:

1) We can have abort race conditions where a request is no longer tracked by the frontend, but still not completed in the engine. See #15326 for an attempt to fix this.
2) We can have async scheduling race conditions where a request ID is removed from the output processor and being scheduled while the older request with that ID is still being completed
   by the model runner. See #29355
3) With P/D, a request will continue to be tracked by the prefill engine long after the prefill request has been completed in the frontend, while we wait for the decode side to fetch the KV blocks. See #20139

Let's instead ensure we use a unique request ID internally, even when a client provides a custom request ID. We can do this simply by appending a short random suffix to any request ID provided by the frontend.

A full 32 character random UUID would be overkill as a suffix, so how many random characters would be sufficient? 8 characters gives us 32 bits of entropy, or 16^8 possible prefixes.

Using the collision probability approximation from https://preshing.com/20110504/hash-collision-probabilities:

N = 16^8 and k is the number of generated suffixes, then the probability of collision is (k^2)/(2N), so If a client somehow
caused vLLM to hold 10k requests that reuse the same client-provided ID, then there would be a 1.16% chance of collision:

```
>>> (k**2)/(2*N)
0.011641532182693481
```

That seems (super good enough)[https://hownot2.com/products/hownot2-super-good-enough-t-shirt].

The key changes to support this are:

1. `InputProcessor.process_inputs()` - we add some randomness to the request ID just before creating an `EngineCoreRequest`, and store both the random "internal" request ID (as `request_id`) and the supplied "external" request ID (as `external_req_id`) in the `EngineCoreRequest`.
2. `RequestState.make_request_output()` - we ensure that `RequestOutput.request_id` continues to be the external request ID (for backwards compat) 
3. `OutputProcessor.abort_requests()` - we make `OutputProcessor` track a mapping from external request ID to internal request IDs, so `abort_requests()` can abort based on either ID.
4. `AsyncLLM` - we use `RequestOutputCollector` to track the internal request ID, so we can use the internal ID to abort an in-progress request. We also add an `internal` boolean flag to `abort()` so API users can abort based on either ID.
5. `ParentRequest` - in the case of parallel sampling, we need to track both the internal and external ID for the later creation of `RequestOutput` aggregating the child outputs.

We need to ensure we track the external->internal request ID mapping because `abort()` will be supplied an external request ID. In the case where an external request ID maps to multiple running requests, we assume the caller requires all of those requests to be aborted. The caller can use `EngineCoreRequest.request_id` or `RequestOutput.internal_request_id` as the request ID (with `internal=True`) if they want to be more specific.

markmc committed:
[Core] Add a random suffix to frontend-provided request IDs

Since #9550 and #10968 we support client's supplying a custom
request ID. The motivation for this is that it can be very helpful
when you need to correlate vLLM logs with logs of a related service.

Since the request ID is used ubiquitously across vLLM as a unique
key, it obviously is problematic if we ever have multiple in-flight
requests using the same client-provided request ID.

We saw this happening recently when `vllm serve bench` started
including a request ID and the request IDs from multiple concurrent
instances caused collisions. See #27723

We try to guard against request ID collisions currently in the
frontend in OutputProcessor:

```
    def add_request(...):
        if request_id in self.request_states:
            raise ValueError(f"Request id {request_id} already running.")
```

however, this is not always effective:

1) We can have abort race conditions where a request is no longer
   tracked by the frontend, but still not completed in the engine.
   See #15326 for an attempt to fix this.
2) We can have async scheduling race conditions where a request
   ID is removed from the output processor and being scheduled
   while the older request with that ID is still being completed
   by the model runner. See #29355
3) With P/D, a request will continue to be tracked by the prefill
   engine long after the prefill request has been completed in
   the frontend, while we wait for the decode side to fetch the
   KV blocks. See #20139

Let's instead ensure we use a unique request ID internally, even
when a client provides a custom request ID. We can do this simply
by appending a short random suffix to any request ID provided
by the frontend.

A full 32 character random UUID would be overkill as a suffix,
so how many random characters would be sufficient? 8 characters
gives us 32 bits of entropy, or 16^8 possible prefixes.

Using the collision probability approximation from
https://preshing.com/20110504/hash-collision-probabilities:

N = 16^8 and k is the number of generated suffixes, then the
probability of collision is (k^2)/(2N), so If a client somehow
caused vLLM to hold 10k requests that reuse the same client-provided
ID, then there would be a 1.16% chance of collision:

```
>>> (k**2)/(2*N)
0.011641532182693481
```

That seems (super good enough)[https://hownot2.com/products/hownot2-super-good-enough-t-shirt].

The key changes to support this are:

1. `InputProcessor.process_inputs()` - we add some randomness to the
request ID just before creating an `EngineCoreRequest`, and store both
the random "internal" request ID (as `request_id`) and the supplied
"external" request ID (as `external_req_id`) in the
`EngineCoreRequest`.
2. `RequestState.make_request_output()` - we ensure that
`RequestOutput.request_id` continues to be the external request ID
(for backwards compat) and add `internal_request_id`.
3. `OutputProcessor.abort_requests()` - we make `OutputProcessor`
track a mapping from external request ID to internal request IDs, so
`abort_requests()` can abort based on either ID.
4. `AsyncLLM` - we use `RequestOutputCollector` to track the internal
request ID, so we can use the internal ID to abort an in-progress
request. We also add an `internal` boolean flag to `abort()` so API
users can abort based on either ID.
5. `ParentRequest` - in the case of parallel sampling, we need to
track both the internal and external ID for the later creation of
`RequestOutput` aggregating the child outputs.

We need to ensure we track the external->internal request ID
mapping because abort() will be supplied an external request ID.
In the case where an external request ID maps to multiple running
requests, we assume the caller requires all of those requests
to be aborted. The caller can use EngineCoreRequest.request_id
as the request ID if they want to be more specific.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -438,6 +458,11 @@ def add_request(
 
+        # Track the external_req_id -> [internal_req_id, ...] mapping
+        if req_state.external_req_id not in self.external_req_ids:
+            self.external_req_ids[req_state.external_req_id] = []
```
Could consider using a `defaultdict(list)` for this, then these lines wouldn't be needed.

njhill commented on vllm/entrypoints/pooling/embed/serving.py:
```diff
@@ -536,8 +536,8 @@ async def _collect_batch(
-                        # Last part should be prompt index
-                        prompt_idx = int(parts[-1])
+                        # Second-to-last part should be prompt index
+                        prompt_idx = int(parts[-2])
```
Yuck, the way the request id is abused in this file is pretty bad. What if an external request id is sent in that happens to contains `-chunk-` or whatever.

(not related to your changes, just hadn't seen this code before!)

markmc replied:
Agree!

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -378,8 +382,24 @@ def abort_requests(
         self,
         request_ids: Iterable[str],
```
We could also consider adding an `internal: bool = False` arg here to allow caller to specify whether or not the ids are internal or external. This might be a bit more robust, for all the places we call abort internally, we should be able to use the internal id and set it to `True`, and external usage shouldn't be affected.

njhill reviewed (commented):
Thanks very much for this @markmc!

Some thoughts in addition to inline comments:
- I was thinking in `AsyncLLM` we should have the `add_request` method return the internal request id as well as the output collector, and then use this in the aborts in the except blocks in `generate()`
- Wondering whether we should set the request id in the outputs from the engine to be the external id (if provided). This would be more backwards compatible and makes more sense to me logically. It also means we wouldn't need the new sorting "hack" in llm.py

markmc commented on vllm/v1/engine/input_processor.py:
```diff
@@ -382,6 +382,12 @@ def _extract_mm_data(p: PromptType):
+        """Construct an internal request ID by adding 8 random characters
+        to the supplied request ID in order to ensure uniquness.
+        """
+        return f"{request_id}-{random_uuid()[:8]}"
```
Done, thanks

markmc commented on vllm/v1/engine/output_processor.py:
```diff
@@ -378,8 +382,24 @@ def abort_requests(
         for request_id in request_ids:
+            # External ID - abort all requests in the external->internal mapping
+            if request_id in self.external_req_ids:
+                internal_req_ids.extend(self.external_req_ids.pop(request_id))
```
Not a fan of this in terms of readability, but done. Thanks

(Also, needed to pass `None` default to `pop()`)

markmc commented on vllm/v1/engine/output_processor.py:
```diff
@@ -378,8 +382,24 @@ def abort_requests(
+            # External ID - abort all requests in the external->internal mapping
+            if request_id in self.external_req_ids:
+                internal_req_ids.extend(self.external_req_ids.pop(request_id))
+                continue
```
If the request ID isn't tracked as an external ID, it is an internal ID

If it is not in `request_states`, then we have no way of mapping this internal ID to the external one

Or, put another way - the code beyond the continue requires the `request_states` lookup to succeed?

In the code suggestion below, you seem to be proposing moving from:

```
if request_id not in self.request_states:
    continue
internal_req_ids.append(request_id)
[...]    
```

to:

```
req_state = self.request_states.get(request_id)
if req_state is not None:
    internal_req_ids.append(request_id)
    [...]    
```

Which is equivalent?

Just spelling this out because I might well be missing your point

njhill replied:
I just meant here in theory someone could create a new request with an external id that matches an existing internal id.. (which would then get it's own internal id). However if we get rid of the `internal=None` case then this comment should be n/a.

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -377,9 +387,60 @@ def propagate_error(self, e: Exception):
+        if internal is not None:
+            check_internal, check_external = internal, not internal
+        else:
+            check_internal, check_external = True, True
```
Do we need to support `None` case? Thought it would be a bit simpler if we dont...

njhill replied:
bumping this question

markmc replied:
Good point, dropping that now (I had locked myself into assuming we needed it for the default, backwards compat case). Thanks

njhill commented:
> > > > > * I was thinking in `AsyncLLM` we should have the `add_request` method return the internal request id as well as the output collector, and then use this in the aborts in the except blocks in `generate()`
> > > > 
> > > > 
> > > > Yeah, `RequestOutputCollector` has this ID, I've used it there
> > > 
> > > 
> > > Doh ... it only has it when it has output! So, that's not going to work. Will look for an alternative
> > 
> > 
> > I think it would make sense to have it just return `tuple[RequestOutputCollector, str]`?
> 
> What about API compatibility? (Do you not consider this public?)

I'm actually not sure about this. I'm pretty sure it would only be used via the `generate()/encode()` methods. I think it's unlikely folks would be using it directly. If we are concerned, we could add bool `return_id` arg to control what's returned.

markmc commented:
Ok, the changes are quite scattered, so here's a condensed summary of where we're at now

1. `InputProcessor.process_inputs()` - we add some randomness to the request ID just before creating an `EngineCoreRequest`, and store both the random "internal" request ID (as `request_id`) and the supplied "external" request ID (as `external_req_id`) in the `EngineCoreRequest`.
2. `RequestState.make_request_output()` - we ensure that `RequestOutput.request_id` continues to be the external request ID (for backwards compat) and add `internal_request_id`.
3. `OutputProcessor.abort_requests()` - we make `OutputProcessor` track a mapping from external request ID to internal request IDs, so `abort_requests()` can abort based on either ID.
4. `AsyncLLM` - we use `RequestOutputCollector` to track the internal request ID, so we can use the internal ID to abort an in-progress request. We also add an `internal` boolean flag to `abort()` so API users can abort based on either ID.
5. `ParentRequest` - in the case of parallel sampling, we need to track both the internal and external ID for the later creation of `RequestOutput` aggregating the child outputs.

njhill commented on vllm/v1/engine/output_processor.py:
```diff
@@ -39,13 +40,16 @@ class RequestOutputCollector:
     def put(self, output: RequestOutput | PoolingRequestOutput | Exception) -> None:
         """Non-blocking put operation."""
+        if isinstance(output, (RequestOutput, PoolingRequestOutput)):
+            assert output.internal_req_id == self.request_id
```
could put this in the corresponding elif blocks below to save an additional if (assuming we keep internal_req_id in the outputs, per other comment)

markmc replied:
The next line makes it a bit tricky:

```
        if self.output is None or isinstance(output, Exception):
            self.output = output
```

The assertion wouldn't apply to the `Exception` case. That's why I did it here

njhill replied:
right, I meant the same assert could be included in both of the `elif` blocks below that one

njhill reviewed (commented):
Thanks @markmc, I think it's looking pretty good.

njhill commented:
From the CI, looks like there's another test update to needed to `test_output_processor.py`. We should run the "full" CI on this one too to flush out anything else. And I guess should queue it behind https://github.com/vllm-project/vllm/pull/29665 so as not to break P/D.

I still think it's worth looking at making the provided request id optional as a follow-on.

markmc commented:
Ok, the `RequestOutput.internal_req_id` removal will have to wait until tomorrow. Hopefully what I've pushed now passes CI (I've tried to do a sweep of relevant tests locally) and I think resolves pretty much everything but the `RequestOutput.internal_req_id` thing

markmc commented:
> I still think it's worth looking at making the provided request id optional as a follow-on.

Filed as #30138

njhill reviewed (commented):
Thanks @markmc it's looking great, just some fairly minor remaining thoughts.

markmc commented:
I presume the `tp_config_sweep_accuracy_test.sh` failure:

> Potentially invalid KV blocks for unrecognized request cmpl-4fd0bde0-eb78-4c6a-94e9-40ba29e6645a-0-a8e879dc were retrieved by a decode worker

will be resolved by rebasing onto #29665

njhill reviewed (approved):
I am pretty sure the blackwell and distributed h200 test failures are unrelated but will confirm before we merge this.

markmc committed:
[Core] Defer allocation of internal request IDs

To guard against the possibility of the external caller allocating
an EngineCoreRequest (note - this should be done by InputProcessor)
we defer the allocation of an internal request ID until the request
is added to the engine.

Multi-modal UUIDs seem to be the one place that the request ID is
used before this point, but the only requirement is for the mm UUID
to guarantee uniqueness, so it does not need to match the internal
request ID.

Signed-off-by: Mark McLoughlin <markmc@redhat.com>

njhill commented on vllm/v1/engine/input_processor.py:
```diff
@@ -403,14 +403,19 @@ def _extract_mm_data(p: PromptType):
+        that adds 8 random characters in order to ensure uniquness.
         """
-        return f"{request_id}-{random_uuid():.8}"
+        assert request.external_req_id is None
```
This should probably be changed to raise a runtime exception. It should be ok to fail the request in this case rather than crash the server.

njhill commented on vllm/v1/engine/input_processor.py:
```diff
@@ -403,9 +403,20 @@ def _extract_mm_data(p: PromptType):
+            ]
         return mm_uuids
 
+    def assign_request_id(self, request: EngineCoreRequest):
```
could make this a `@staticmethod` to make it clear that it's stateless

njhill reviewed (commented):
Thanks @markmc! Looking forward to landing this!

njhill committed:
Updates based on review

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fix test

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge branch 'main' into random-request-id

ywang96 reviewed (approved):
LGTM - not particular concern from me

njhill committed:
Merge branch 'main' into random-request-id

njhill committed:
Merge branch 'main' into random-request-id

njhill committed:
Merge branch 'main' into random-request-id

njhill committed:
Merge branch 'main' into random-request-id

njhill committed:
Merge branch 'main' into random-request-id
